Tasks/ServiceFabricDeploy/Update-DockerSettings.psm1

```diff
@@ -0,0 +1,118 @@
+            $serviceManifestXml = [xml](Get-Content $serviceManifestPath)
+
+            # Enumerate through the code packages of the service manifest and find those that are associated with an image name. For each
+            # of those code packages, we're going to setup the app manifest with ContainerHostPolicies for those code packages so that they
+            # are configured to use the selected registry's credentials.  Example XML fragment of what we're trying to ensure exists in
+            # the app manifest:
+            #   <ServiceManifestImport>
+            #     <ServiceManifestRef ServiceManifestName="WebApplication1Pkg" ServiceManifestVersion="1.0.0" />
+            #     <Policies>
+            #       <ContainerHostPolicies CodePackageRef="Code">
+            #         <RepositoryCredentials AccountName="username" Password="encryptedpassword" PasswordEncrypted="true" />
+            #       </ContainerHostPolicies>
+            #     </Policies>
+            #   </ServiceManifestImport>
+
+            $codePackages = @($serviceManifestXml.ServiceManifest.CodePackage)
+            foreach ($codePackage in $codePackages)
+            {
+                # If this code package has an image name associated with it, update the app manifest to ensure there are respository credentials defined for this package.
+                if ($codePackage.EntryPoint -and $codePackage.EntryPoint.ContainerHost -and $codePackage.EntryPoint.ContainerHost.ImageName)
+                {
+                    # Ensure there is a Policies element
+                    $policies = $serviceManifestImport.Policies
+                    if (-not $serviceManifestImport.Policies)
+                    {
+                        $policies = $appManifestXml.CreateElement("Policies", $appManifestXml.ApplicationManifest.NamespaceURI)
+                        [void]($serviceManifestImport.AppendChild($policies))
+                    }
+
+                    # Ensure there is a ContainerHostPolicies element
+                    $containerHostPolicies = @($policies.ContainerHostPolicies)
+                    $containerHostPolicy = $containerHostPolicies | where { $_.CodePackageRef -eq $codePackage.Name }
+                    if (-not $containerHostPolicy)
+                    {
+                        $containerHostPolicy = $appManifestXml.CreateElement("ContainerHostPolicies", $appManifestXml.ApplicationManifest.NamespaceURI)
+                        $containerHostPolicy.SetAttribute("CodePackageRef", $codePackage.Name)
+                        [void]($policies.AppendChild($containerHostPolicy))
+                    }
+
+                    # If there isn't already a RepositoryCredentials element, create one with the credentials of the selected registry
+                    if (-not $containerHostPolicy.RepositoryCredentials)
```

Comment:
shouldn't existing credential be overwritten?

Reply:
It's possible the customer setup their own container services outside the context of our tooling, in which case they would be specifying these credentials explicitly in the manifest.  We wouldn't want to overwrite those settings.

---
In reply to: [137857180](https://github.com/Microsoft/vsts-tasks/pull/5076#discussion_r137857180) [](ancestors = 137857180)